src/test/java/uk/gov/cshr/controller/emailUpdate/AgencyTokenVerificationControllerTest.java

```diff
@@ -0,0 +1,197 @@
+                .andExpect(model().attribute("domain", DOMAIN))
+                .andExpect(model().attribute("uid", IDENTITY_UID))
+                .andExpect(model().attribute("code", CODE))
+                .andExpect(model().attributeExists("emailUpdatedRecentlyEnterTokenForm"))
+                .andExpect(view().name("enterTokenSinceEmailUpdate"));
+    }
+
+    @Test
+    public void givenAValidTokenForm_whenCheckToken_thenShouldSubmitToken() throws Exception {
+        AgencyToken agencyToken = new AgencyToken();
+        agencyToken.setUid(AGENCY_TOKEN_UID);
+
+        when(csrsService.getAgencyTokenForDomainTokenOrganisation(DOMAIN, TOKEN, ORGANISATION)).thenReturn(Optional.of(agencyToken));
+        when(agencyTokenCapacityService.hasSpaceAvailable(agencyToken)).thenReturn(true);
+
+        EmailUpdate emailUpdate = new EmailUpdate();
+        emailUpdate.setEmail(EMAIL);
+        Identity identity = new Identity();
+        identity.setEmail(EMAIL);
+        when(emailUpdateService.getEmailUpdate(any(Identity.class), eq(CODE))).thenReturn(emailUpdate);
+        doNothing().when(emailUpdateService).updateEmailAddress(any(HttpServletRequest.class), eq(identity), eq(emailUpdate), eq(agencyToken));
+
+        mockMvc.perform(
+                post(VERIFY_TOKEN_URL + CODE)
+                        .with(CsrfRequestPostProcessor.csrf())
+                        .param("organisation", ORGANISATION)
+                        .param("token", TOKEN)
+                        .param("domain", DOMAIN)
+                        .param("uid", IDENTITY_UID)
+        )
+                .andExpect(status().is3xxRedirection())
+                .andExpect(redirectedUrl(REDIRECT_EMAIL_UPDATED));
+    }
+
+    @Test
+    public void givenAValidTokenForm_whenCheckTokenAndNoSpacesAvailable_thenShouldRedirect() throws Exception {
+        AgencyToken agencyToken = new AgencyToken();
+        agencyToken.setUid(AGENCY_TOKEN_UID);
+
+        when(csrsService.getAgencyTokenForDomainTokenOrganisation(DOMAIN, TOKEN, ORGANISATION)).thenReturn(Optional.of(agencyToken));
+        when(agencyTokenCapacityService.hasSpaceAvailable(agencyToken)).thenReturn(false);
+
+        EmailUpdate emailUpdate = new EmailUpdate();
+        Identity identity = new Identity();
+        when(emailUpdateService.getEmailUpdate(identity, CODE)).thenReturn(emailUpdate);
+        doNothing().when(emailUpdateService).updateEmailAddress(any(HttpServletRequest.class), eq(identity), eq(emailUpdate), eq(agencyToken));
+
+        mockMvc.perform(
+                post(VERIFY_TOKEN_URL + CODE)
+                        .with(CsrfRequestPostProcessor.csrf())
+                        .param("organisation", ORGANISATION)
+                        .param("token", TOKEN)
+                        .param("domain", DOMAIN)
+                        .param("uid", IDENTITY_UID)
+        )
+                .andExpect(status().is3xxRedirection())
```

Comment:
This unit test should also have checks for the following:

1. The status attribute is in the model with the correct error message (not covered)
2. The form is the model (not covered)
3. the redirected url is as expected (already covered)